File: BydStats/Models/ChargingRatePlan.swift
import Foundation

// ─── 계절 ───────────────────────────────────────────────────────────────────

enum Season {
    case summer      // 6-8월
    case springFall  // 3-5월, 9-10월
    case winter      // 11-2월

    static func of(month: Int) -> Season {
        switch month {
        case 6, 7, 8:       return .summer
        case 11, 12, 1, 2:  return .winter
        default:             return .springFall
        }
    }
}

// ─── 시간대 슬롯 ─────────────────────────────────────────────────────────────

struct RateSlot: Identifiable {
    let id = UUID()
    let startHour: Int   // inclusive
    let endHour: Int     // exclusive
    let rate: Double     // 원/kWh
    let label: String

    var timeString: String {
        String(format: "%02d:00~%02d:00", startHour, endHour)
    }

    func contains(hour: Int) -> Bool {
        startHour < endHour ? (startHour..<endHour).contains(hour)
                            : hour >= startHour || hour < endHour
    }
}

// ─── 요금제 ──────────────────────────────────────────────────────────────────

enum ChargingRatePlan: Equatable {
    case flat(id: String, name: String, provider: String, rate: Double)
    case timeOfUse(id: String, name: String, provider: String, slots: [Season: [RateSlot]])
    case custom(rate: Double)

    var id: String {
        switch self {
        case .flat(let id, _, _, _):       return id
        case .timeOfUse(let id, _, _, _):  return id
        case .custom:                      return "custom"
        }
    }

    var name: String {
        switch self {
        case .flat(_, let name, _, _):       return name
        case .timeOfUse(_, let name, _, _):  return name
        case .custom:                        return "직접 입력"
        }
    }

    var provider: String {
        switch self {
        case .flat(_, _, let p, _):       return p
        case .timeOfUse(_, _, let p, _):  return p
        case .custom:                     return "기타"
        }
    }

    /// 해당 시점의 요금 (원/kWh)
    func rate(at date: Date) -> Double {
        switch self {
        case .flat(_, _, _, let r):           return r
        case .custom(let r):                  return r
        case .timeOfUse(_, _, _, let slots):
            let cal = Calendar.current
            let month = cal.component(.month, from: date)
            let hour  = cal.component(.hour, from: date)
            let season = Season.of(month: month)
            let s = slots[season] ?? []
            return s.first(where: { $0.contains(hour: hour) })?.rate ?? s.first?.rate ?? 0
        }
    }

    /// 현재 계절 기준 슬롯 (설정 화면 표시용)
    func currentSlots() -> [RateSlot] {
        switch self {
        case .flat(_, _, _, let r):
            return [RateSlot(startHour: 0, endHour: 24, rate: r, label: "단일 단가")]
        case .custom(let r):
            return [RateSlot(startHour: 0, endHour: 24, rate: r, label: "단일 단가")]
        case .timeOfUse(_, _, _, let slots):
            let month = Calendar.current.component(.month, from: Date())
            return slots[Season.of(month: month)] ?? []
        }
    }

    static func == (lhs: ChargingRatePlan, rhs: ChargingRatePlan) -> Bool {
        lhs.id == rhs.id
    }
}

// ─── 사전 정의 요금제 ─────────────────────────────────────────────────────────
// 출처: 한국전력 전기자동차 충전전력요금(자가소비용), 2026.4.16 시행

private let kepcoLowSlots: [Season: [RateSlot]] = [
    .summer: [
        RateSlot(startHour:  0, endHour:  8, rate:  84.3, label: "경부하"),
        RateSlot(startHour:  8, endHour: 15, rate: 172.0, label: "중간부하"),
        RateSlot(startHour: 15, endHour: 21, rate: 259.2, label: "최대부하"),
        RateSlot(startHour: 21, endHour: 22, rate: 172.0, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate:  84.3, label: "경부하"),
    ],
    .springFall: [
        RateSlot(startHour:  0, endHour:  8, rate:  85.4, label: "경부하"),
        RateSlot(startHour:  8, endHour: 15, rate:  97.2, label: "중간부하"),
        RateSlot(startHour: 15, endHour: 21, rate: 102.1, label: "최대부하"),
        RateSlot(startHour: 21, endHour: 22, rate:  97.2, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate:  85.4, label: "경부하"),
    ],
    .winter: [
        RateSlot(startHour:  0, endHour:  8, rate: 107.4, label: "경부하"),
        RateSlot(startHour:  8, endHour:  9, rate: 154.9, label: "중간부하"),
        RateSlot(startHour:  9, endHour: 12, rate: 217.5, label: "최대부하"),
        RateSlot(startHour: 12, endHour: 16, rate: 154.9, label: "중간부하"),
        RateSlot(startHour: 16, endHour: 19, rate: 217.5, label: "최대부하"),
        RateSlot(startHour: 19, endHour: 22, rate: 154.9, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate: 107.4, label: "경부하"),
    ],
]

private let kepcoHighSlots: [Season: [RateSlot]] = [
    .summer: [
        RateSlot(startHour:  0, endHour:  8, rate:  79.2, label: "경부하"),
        RateSlot(startHour:  8, endHour: 15, rate: 137.4, label: "중간부하"),
        RateSlot(startHour: 15, endHour: 21, rate: 190.4, label: "최대부하"),
        RateSlot(startHour: 21, endHour: 22, rate: 137.4, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate:  79.2, label: "경부하"),
    ],
    .springFall: [
        RateSlot(startHour:  0, endHour:  8, rate:  80.2, label: "경부하"),
        RateSlot(startHour:  8, endHour: 15, rate:  91.0, label: "중간부하"),
        RateSlot(startHour: 15, endHour: 21, rate:  94.9, label: "최대부하"),
        RateSlot(startHour: 21, endHour: 22, rate:  91.0, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate:  80.2, label: "경부하"),
    ],
    .winter: [
        RateSlot(startHour:  0, endHour:  8, rate:  96.6, label: "경부하"),
        RateSlot(startHour:  8, endHour:  9, rate: 127.7, label: "중간부하"),
        RateSlot(startHour:  9, endHour: 12, rate: 165.5, label: "최대부하"),
        RateSlot(startHour: 12, endHour: 16, rate: 127.7, label: "중간부하"),
        RateSlot(startHour: 16, endHour: 19, rate: 165.5, label: "최대부하"),
        RateSlot(startHour: 19, endHour: 22, rate: 127.7, label: "중간부하"),
        RateSlot(startHour: 22, endHour: 24, rate:  96.6, label: "경부하"),
    ],
]

let predefinedRatePlans: [ChargingRatePlan] = [
    .timeOfUse(id: "kepco_low",  name: "한전 비공용 저압", provider: "한국전력 (가정용 완속)", slots: kepcoLowSlots),
    .timeOfUse(id: "kepco_high", name: "한전 비공용 고압", provider: "한국전력 (상업용 고압)", slots: kepcoHighSlots),
]

func ratePlan(id: String, customRate: Double = 180.0) -> ChargingRatePlan {
    predefinedRatePlans.first(where: { $0.id == id }) ?? .custom(rate: customRate)
}

File: androidApp/src/main/kotlin/com/ggpark/bydstats/android/data/ChargingRatePlan.kt
package com.ggpark.bydstats.android.data

import java.util.Calendar

// ─── 계절 ───────────────────────────────────────────────────────────────────

enum class Season {
    SUMMER,      // 6-8월
    SPRING_FALL, // 3-5월, 9-10월
    WINTER;      // 11-2월

    companion object {
        fun of(month: Int): Season = when (month) {
            6, 7, 8       -> SUMMER
            11, 12, 1, 2  -> WINTER
            else          -> SPRING_FALL
        }
    }
}

// ─── 시간대 슬롯 ─────────────────────────────────────────────────────────────

data class RateSlot(
    val startHour: Int,  // inclusive
    val endHour: Int,    // exclusive (24 = 자정)
    val rate: Double,    // 원/kWh
    val label: String,
) {
    /** hour가 이 슬롯에 포함되는지 확인 */
    fun contains(hour: Int): Boolean =
        if (startHour < endHour) hour in startHour until endHour
        else hour >= startHour || hour < endHour  // 자정 넘기는 구간 (예: 22-08)
}

// ─── 요금제 ──────────────────────────────────────────────────────────────────

sealed class ChargingRatePlan {
    abstract val id: String
    abstract val name: String
    abstract val provider: String

    /** 해당 epoch ms 시점의 요금(원/kWh) */
    abstract fun rateAt(timestampMs: Long): Double

    /** 설정 화면 표시용: 현재 계절 기준 슬롯 */
    abstract fun currentSlots(): List<RateSlot>

    // 단일 단가
    data class Flat(
        override val id: String,
        override val name: String,
        override val provider: String,
        val rate: Double,
    ) : ChargingRatePlan() {
        override fun rateAt(timestampMs: Long) = rate
        override fun currentSlots() = listOf(RateSlot(0, 24, rate, "단일 단가"))
    }

    // 계절별 시간대 차등
    data class TimeOfUse(
        override val id: String,
        override val name: String,
        override val provider: String,
        val seasonalSlots: Map<Season, List<RateSlot>>,
    ) : ChargingRatePlan() {
        override fun rateAt(timestampMs: Long): Double {
            val cal = Calendar.getInstance().apply { timeInMillis = timestampMs }
            val month = cal.get(Calendar.MONTH) + 1
            val hour  = cal.get(Calendar.HOUR_OF_DAY)
            val slots = seasonalSlots[Season.of(month)] ?: return 0.0
            return slots.firstOrNull { it.contains(hour) }?.rate ?: slots.first().rate
        }

        override fun currentSlots(): List<RateSlot> {
            val month = Calendar.getInstance().get(Calendar.MONTH) + 1
            return seasonalSlots[Season.of(month)] ?: emptyList()
        }
    }

    // 직접 입력
    class Custom(val rate: Double) : ChargingRatePlan() {
        override val id = "custom"
        override val name = "직접 입력"
        override val provider = "기타"
        override fun rateAt(timestampMs: Long) = rate
        override fun currentSlots() = listOf(RateSlot(0, 24, rate, "단일 단가"))
    }
}

// ─── 사전 정의 요금제 ─────────────────────────────────────────────────────────
// 출처: 한국전력 전기자동차 충전전력요금(자가소비용), 2026.4.16 시행

// 한전 비공용 저압 (가정용 비공용 완속)
private val KEPCO_LOW_SLOTS = mapOf(
    Season.SUMMER to listOf(
        RateSlot(0, 8,  84.3, "경부하"),   // 00~08
        RateSlot(8, 15, 172.0, "중간부하"), // 08~15
        RateSlot(15, 21, 259.2, "최대부하"), // 15~21
        RateSlot(21, 22, 172.0, "중간부하"), // 21~22
        RateSlot(22, 24, 84.3, "경부하"),   // 22~24
    ),
    Season.SPRING_FALL to listOf(
        RateSlot(0, 8,  85.4, "경부하"),
        RateSlot(8, 15, 97.2, "중간부하"),
        RateSlot(15, 21, 102.1, "최대부하"),
        RateSlot(21, 22, 97.2, "중간부하"),
        RateSlot(22, 24, 85.4, "경부하"),
    ),
    Season.WINTER to listOf(
        RateSlot(0, 8,  107.4, "경부하"),
        RateSlot(8, 9,  154.9, "중간부하"), // 08~09
        RateSlot(9, 12, 217.5, "최대부하"), // 09~12
        RateSlot(12, 16, 154.9, "중간부하"), // 12~16
        RateSlot(16, 19, 217.5, "최대부하"), // 16~19
        RateSlot(19, 22, 154.9, "중간부하"), // 19~22
        RateSlot(22, 24, 107.4, "경부하"),
    ),
)

// 한전 비공용 고압
private val KEPCO_HIGH_SLOTS = mapOf(
    Season.SUMMER to listOf(
        RateSlot(0, 8,  79.2, "경부하"),
        RateSlot(8, 15, 137.4, "중간부하"),
        RateSlot(15, 21, 190.4, "최대부하"),
        RateSlot(21, 22, 137.4, "중간부하"),
        RateSlot(22, 24, 79.2, "경부하"),
    ),
    Season.SPRING_FALL to listOf(
        RateSlot(0, 8,  80.2, "경부하"),
        RateSlot(8, 15, 91.0, "중간부하"),
        RateSlot(15, 21, 94.9, "최대부하"),
        RateSlot(21, 22, 91.0, "중간부하"),
        RateSlot(22, 24, 80.2, "경부하"),
    ),
    Season.WINTER to listOf(
        RateSlot(0, 8,  96.6, "경부하"),
        RateSlot(8, 9,  127.7, "중간부하"),
        RateSlot(9, 12, 165.5, "최대부하"),
        RateSlot(12, 16, 127.7, "중간부하"),
        RateSlot(16, 19, 165.5, "최대부하"),
        RateSlot(19, 22, 127.7, "중간부하"),
        RateSlot(22, 24, 96.6, "경부하"),
    ),
)

val PREDEFINED_RATE_PLANS: List<ChargingRatePlan> = listOf(
    ChargingRatePlan.TimeOfUse(
        id = "kepco_low",
        name = "한전 비공용 저압",
        provider = "한국전력 (가정용 완속)",
        seasonalSlots = KEPCO_LOW_SLOTS,
    ),
    ChargingRatePlan.TimeOfUse(
        id = "kepco_high",
        name = "한전 비공용 고압",
        provider = "한국전력 (상업용 고압)",
        seasonalSlots = KEPCO_HIGH_SLOTS,
    ),
)

fun ratePlanById(id: String, customRate: Double = 180.0): ChargingRatePlan =
    PREDEFINED_RATE_PLANS.firstOrNull { it.id == id } ?: ChargingRatePlan.Custom(customRate)
